packages/modelence/src/lock/helpers.ts

```diff
@@ -0,0 +1,148 @@
+ * If the lock already exists and is owned by another container, this will fail.
+ * If the lock is expired (no recent heartbeat), it will be taken over.
+ *
+ * @param resource - The type of lock to acquire
+ * @param options.lockDuration - Time in ms after which a lock is considered expired (default: 30s)
+ * @param options.lockCacheDuration - Time in ms to cache successful lock acquisition (default: 10s)
+ * @param options.instanceId - The unique identifier for this application instance
+ * @returns true if lock was acquired, false otherwise
+ */
+export async function acquireLock(
+  resource: string,
+  {
+    lockDuration = DEFAULT_LOCK_DURATION,
+    successfulLockCacheDuration = DEFAULT_CACHE_DURATION,
+    failedLockCacheDuration = DEFAULT_CACHE_DURATION,
+    instanceId = INSTANCE_ID,
+  }: {
+    lockDuration?: number;
+    successfulLockCacheDuration?: number;
+    failedLockCacheDuration?: number;
+    instanceId?: string;
+  } = {}
+): Promise<boolean> {
+  const now = Date.now();
+  if (lockCache[resource] && now - lockCache[resource].expiresAt < 0) {
+    return lockCache[resource].value;
+  }
+
+  const staleThresholdDate = new Date(now - lockDuration);
+
+  logDebug(`Attempting to acquire lock: ${resource}`, {
+    source: 'lock',
+    resource,
+    instanceId,
+  });
+
+  try {
+    const result = await locksCollection.upsertOne(
+      {
+        $or: [
+          { resource, instanceId },
+          { resource, acquiredAt: { $lt: staleThresholdDate } },
+        ],
+      },
+      {
+        $set: {
+          resource,
+          instanceId,
+          acquiredAt: new Date(),
+        },
+      }
+    );
+
+    const isLockAcquired = result.upsertedCount > 0 || result.modifiedCount > 0;
+
+    lockCache[resource] = {
+      value: isLockAcquired,
+      expiresAt: now + (isLockAcquired ? successfulLockCacheDuration : failedLockCacheDuration),
+    };
+
+    if (isLockAcquired) {
+      logDebug(`Lock acquired: ${resource}`, {
+        source: 'lock',
+        resource,
+        instanceId,
+      });
+    } else {
+      logDebug(`Failed to acquire lock (already held): ${resource}`, {
+        source: 'lock',
+        resource,
+        instanceId,
+      });
+    }
+
+    return isLockAcquired;
+  } catch {
+    lockCache[resource] = {
+      value: false,
+      expiresAt: now + failedLockCacheDuration,
+    };
+    logDebug(`Failed to acquire lock (already held): ${resource}`, {
+      source: 'lock',
+      resource,
+      instanceId,
+    });
+    return false;
+  }
+}
+
+/**
+ * Releases a lock of the specified resource owned by this container.
+ *
+ * @param resource - The resource to release the lock for
+ * @param options.instanceId - The unique identifier for this application instance
+ * @returns true if lock was released, false if lock wasn't owned by this container
+ */
+export async function releaseLock(
+  resource: string,
+  {
+    instanceId = INSTANCE_ID,
+  }: {
+    instanceId?: string;
```

Comment:
Shouldn't this be a non-optional parameter? Otherwise if it's undefined, the `deleteOne` can have unintended consequence and will delete whatever lock there is, even if it's not owned by this instance.

Reply:
Sure, functionally I think it's good, but shouldn't the type be `instanceId: string` then without `?` ?